lib/ui/painting/display_list_deferred_image_gpu.cc

```diff
@@ -28,43 +57,132 @@ std::shared_ptr<impeller::Texture> DlDeferredImageGPU::impeller_texture()
   }
-  return sizeof(this) + size_.width() * size_.height() * 4;
+  return sizeof(this);
 }
 
-void DlDeferredImageGPU::set_image(sk_sp<SkImage> image) {
-  FML_DCHECK(image);
-  FML_DCHECK(image->dimensions() == size_);
-  image_ = std::move(image);
+std::optional<std::string> DlDeferredImageGPU::get_error() const {
+  if (auto image_wrapper = std::atomic_load(&image_wrapper_)) {
+    return image_wrapper->get_error();
+  }
+  return std::nullopt;
 }
 
-void DlDeferredImageGPU::set_error(const std::string& error) {
-  FML_DCHECK(!image_);
-  std::scoped_lock lock(error_mutex_);
-  error_ = std::move(error);
+std::shared_ptr<DlDeferredImageGPU::ImageWrapper>
+DlDeferredImageGPU::ImageWrapper::Make(
+    const SkImageInfo& image_info,
+    sk_sp<DisplayList> display_list,
+    fml::WeakPtr<SnapshotDelegate> snapshot_delegate,
+    fml::RefPtr<fml::TaskRunner> raster_task_runner,
+    fml::RefPtr<SkiaUnrefQueue> unref_queue) {
+  auto wrapper = std::shared_ptr<ImageWrapper>(new ImageWrapper(
+      image_info, std::move(display_list), std::move(snapshot_delegate),
+      std::move(raster_task_runner), std::move(unref_queue)));
+  wrapper->SnapshotDisplayList();
+  return wrapper;
 }
 
-std::optional<std::string> DlDeferredImageGPU::get_error() const {
+DlDeferredImageGPU::ImageWrapper::ImageWrapper(
+    const SkImageInfo& image_info,
+    sk_sp<DisplayList> display_list,
+    fml::WeakPtr<SnapshotDelegate> snapshot_delegate,
+    fml::RefPtr<fml::TaskRunner> raster_task_runner,
+    fml::RefPtr<SkiaUnrefQueue> unref_queue)
+    : image_info_(image_info),
+      display_list_(std::move(display_list)),
+      snapshot_delegate_(std::move(snapshot_delegate)),
+      raster_task_runner_(std::move(raster_task_runner)),
+      unref_queue_(std::move(unref_queue)) {}
+
+void DlDeferredImageGPU::ImageWrapper::OnGrContextCreated() {
+  FML_DCHECK(raster_task_runner_->RunsTasksOnCurrentThread());
+  SnapshotDisplayList();
+}
+
+void DlDeferredImageGPU::ImageWrapper::OnGrContextDestroyed() {
+  FML_DCHECK(raster_task_runner_->RunsTasksOnCurrentThread());
+
+  DeleteTexture();
+}
+
+sk_sp<SkImage> DlDeferredImageGPU::ImageWrapper::CreateSkiaImage() const {
+  FML_DCHECK(raster_task_runner_->RunsTasksOnCurrentThread());
+
+  if (texture_.isValid() && context_) {
+    return SkImage::MakeFromTexture(
+        context_.get(), texture_, kTopLeft_GrSurfaceOrigin,
+        image_info_.colorType(), image_info_.alphaType(),
+        image_info_.refColorSpace());
+  }
+  return image_;
+}
+
+bool DlDeferredImageGPU::ImageWrapper::isTextureBacked() const {
+  return texture_.isValid();
+}
+
+void DlDeferredImageGPU::ImageWrapper::SnapshotDisplayList() {
+  fml::TaskRunner::RunNowOrPostTask(
+      raster_task_runner_, [weak_this = weak_from_this()]() {
+        auto wrapper = weak_this.lock();
+        if (!wrapper) {
+          return;
+        }
+        auto result = wrapper->snapshot_delegate_->MakeGpuImage(
```

Comment:
Check the validity of the snapshot delegate `fml::WeakPtr` before dereferencing

Reply:
Done